### ISSUE TITLE
[processor/resourcedetection], [receiver/dockerstats] Collector cannot query Docker socket in official contrib images

### ISSUE THREAD

**Issue by mx-psi** [open-telemetry/opentelemetry-collector-contrib#11791]
**Describe the bug**

The [docker detector](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor#docker-metadata) from the resource detection processor and the [dockerstats receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/dockerstatsreceiver) do not work on official opentelemetry-collector-contrib images, or any other image that runs the Collector under a user other than root. 

**Steps to reproduce**

Run the resource detection processor docker detector or the dockerstats receiver, while mounting the `/var/run/docker.sock` socket:

```
docker run -v /var/run/docker.sock:/var/run/docker.sock:ro -v <mount config here> otel/opentelemetry-collector-contrib
```

**What did you expect to see?**

The Docker detector should add the `host.name` of the host machine, and its operating system.

The Docker stats receiver should produce valid metrics.

**What did you see instead?**

Both components fail because of lack of permissions

**What version did you use?**

Can be reproduced on the latest version, happens since v0.40.0 (more specifically, since #6380).

**What config did you use?**

For both components the default configuration on the README can reproduce this; see e.g. the resource detection processor:

```
processors:
  resourcedetection/docker:
    detectors: [env, docker]
    timeout: 2s
    override: false
```

**Environment**

This happens on every Docker version and every Collector image since v0.40.0

**Additional context**

This happens since #6380, because of a permissions issue: the mounted socket is only readable by root. AFAICT, Docker does not currently allow mounting volumes with permissions for a specific user (see moby/moby#2259), and we can't `chown` the socket at build time, so we have to choose between running as rootless or supporting this. 

This is not a problem on downstream or custom distributions that run as root.

For getting the hostname on the Docker detector, a workaround is to override the OS hostname on the Docker image using something like ` --hostname $(hostname)`. I don't know of a workaround for getting the hosts' operating system or getting the metrics on the dockerstats receiver.

**Comment by mx-psi** [comment]
@open-telemetry/collector-contrib-maintainer I assume we want to keep running under a non-root user. If we don't find a solution that works when not running as root, should the docker detector be deprecated and eventually removed? This would still be useful on downstream distros that run as root, but I don't know if that is a common case.

**Comment by TylerHelmuth** [comment]
pinging @jrcamp @pmm-sumo @Aneurysm9 @dashpole as code owners

**Comment by dashpole** [comment]
Using the docker socket is a really high level of privilege generally, and I agree with it not being the recommended configuration.  Mounting individual files (e.g. /etc/hostname) seems like a better way to get some of the information you are interested in than fetching it from docker.  I haven't looked into it at all, but I wonder if the system detector with a few files mounted readonly would work the same as the docker detector.

**Comment by mx-psi** [comment]
The dockerstatsreceiver also queries the Docker socket and thus suffers from the same problem https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/dockerstatsreceiver#configuration

**Comment by ErvalhouS** [comment]
> The dockerstatsreceiver also queries the Docker socket and thus suffers from the same problem https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/dockerstatsreceiver#configuration

This is also why I'm here

**Comment by mx-psi** [comment]
I think realistically we only have two options here:

- (Option 1) deprecate end eventually remove these components
- (Option 2) tell people to run the Collector as root via e.g. `docker run -u 0` and explain the security implications

I feel like, at least for the dockerstats receiver, option 1 would cause a lot of pain, so I would rather work on option 2.

**Comment by rmfitzpatrick** [comment]
This is a general docker concern and the container user needs to be in the host's docker group:

```bash
$ docker run -v /var/run/docker.sock:/var/run/docker.sock:ro --group-add $(stat -c '%g' /var/run/docker.sock) otel/opentelemetry-collector-contrib <...>
# or if specifying the user:group directly
$ docker run -v /var/run/docker.sock:/var/run/docker.sock:ro --user "some.user:$(stat -c '%g' /var/run/docker.sock)" otel/opentelemetry-collector-contrib <...>
```

**Comment by carlreid** [comment]
@gbbr I see you're making a lot of nice progress relating to the `dockerstats` receiver, but I also just hit this permission issue when trying to mount the `docker.sock`. Did you manage to figure something out that works without needing to run as privileged? Especially since we'd like to take advantage of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22149.

**Comment by gbbr** [comment]
@carlreid on this front specifically no. I am not aware of a better method unfortunately.

**Comment by R-Sommer** [comment]
For docker-compose.yml `group_add` could be added with docker's group ID of the host e.g.:
```
group_add:
  - "998"
```
Double quotes are necessary otherwise this error would occur:
```
* 'group_add[0]' expected type 'string', got unconvertible type 'int', value: '998'
```
Using "docker" instead of its ID result in this error:
```
Error response from daemon: Unable to find group docker: no matching entries in group file
```

**Comment by bwoester** [comment]
Dockerstats receiver supports an endpoint configuration. Could the docker resourcedetection processor not follow the same principle?

**Comment by mx-psi** [comment]
> Dockerstats receiver supports an endpoint configuration. Could the docker resourcedetection processor not follow the same principle?

@bwoester Yes, but I don't understand how that would solve this issue, could you elaborate?

**Comment by bwoester** [comment]
I wouldn't claim it solves the issue (issue being that someone needs to talk to the docker socket and requires enough privileges to do so), but it moves it away from otel and allows to deal with that specific concern in one location.

The idea would be to have a dedicated service up and running that acts as a reverse proxy over docker's REST API.
This service needs the docker socket, but it exposes the REST API available to everyone, no special user or group IDs required.
So on first glance, this makes everything really insecure.
But the service can also restrict access to certain API endpoints, like building or pulling images, creating new containers and so on.
If it's configured to only allow read-access, I think things already become more secure.
Access to the service may further be protected by authentication mechanisms and so on...

See https://github.com/Tecnativa/docker-socket-proxy for an example of such a proxy.

**Comment by jamesmoessis** [comment]
It seems to be the best solution here is to provide more information to users who want to run this. There isn't anything really that mentions permissions in the [docker stats receiver README](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/README.md), for example.

I'm not sure from the OpenTelemetry side we can really do much besides this, if the collector needs access to the docker socket, the user will need to configure extra permissions for this. I think the best we can do is give the best guidance we can via documentation.

In terms of @bwoester and use of the docker socket proxy, this is a solution people can employ outside of OpenTelemetry, if they have concerns around security implications. Another way security issues can be mitigated is isolating components that need elevated privileges into their own process. So running the docker stats receiver in it's own collector, and then everything else somewhere else. All of these solutions require extra effort but I don't think we can solve them from the perspective of OpenTelemetry.

@mx-psi do you agree with this way forward. We can create actions to uplift the documentation in this area. We also had #37849 opened as a similar issue, so this is affecting some users still.

**Comment by mx-psi** [comment]
@jamesmoessis sorry for the delay. That plan sounds reasonable to me 👍
